File: sdk/go/client/payments.go
package client

import (
	"context"

	clientv1 "github.com/torchwooddev/torchwood/genproto/client/v1"
	"google.golang.org/protobuf/types/known/structpb"
)

// PaymentsService 封装 Client API 的支付服务（建单 + 本人订单查询）。
type PaymentsService struct{ c *Client }

// CreateOrder 创建支付订单。amount 为最小货币单位 int64（禁止 float）。
func (p *PaymentsService) CreateOrder(ctx context.Context, req *clientv1.CreateOrderRequest) (*clientv1.CreateOrderResponse, error) {
	return p.c.payments.CreateOrder(ctx, req)
}

// CreateOrderValues 用基本类型建单（purpose 为 JSON 兼容 map）。
func (p *PaymentsService) CreateOrderValues(ctx context.Context, idempotencyKey, provider string, amount int64, currency, purposeKind string, purpose map[string]any) (*clientv1.CreateOrderResponse, error) {
	var purposeStruct *structpb.Struct
	if len(purpose) > 0 {
		st, err := toStruct(purpose)
		if err != nil {
			return nil, err
		}
		purposeStruct = st
	}
	return p.CreateOrder(ctx, &clientv1.CreateOrderRequest{
		IdempotencyKey: idempotencyKey,
		Provider:       provider,
		Amount:         amount,
		Currency:       currency,
		PurposeKind:    purposeKind,
		Purpose:        purposeStruct,
	})
}

// GetMyOrder 查询本人订单。
func (p *PaymentsService) GetMyOrder(ctx context.Context, orderID string) (*clientv1.PaymentOrder, error) {
	return p.c.payments.GetMyOrder(ctx, &clientv1.GetMyOrderRequest{OrderId: orderID})
}

// ListMyOrders 列出本人订单。
func (p *PaymentsService) ListMyOrders(ctx context.Context, pageSize int32, pageToken string) (*clientv1.ListMyOrdersResponse, error) {
	return p.c.payments.ListMyOrders(ctx, &clientv1.ListMyOrdersRequest{
		PageSize:  pageSize,
		PageToken: pageToken,
	})
}

// VerifyReceipt 校验 iOS IAP receipt / StoreKit 2 JWS 并履约对应订单。
func (p *PaymentsService) VerifyReceipt(ctx context.Context, orderID, receipt string) (*clientv1.VerifyReceiptResponse, error) {
	return p.c.payments.VerifyReceipt(ctx, &clientv1.VerifyReceiptRequest{
		OrderId: orderID,
		Receipt: receipt,
	})
}

File: sdk/typescript/src/client/payments.ts
import { listQuery, type HttpTransport } from "../http.js";
import type { CreateOrderResponse, ListParams, PaymentOrder } from "../types.js";

export class ClientPaymentsService {
  constructor(private readonly http: HttpTransport) {}

  async createOrder(input: {
    idempotency_key: string;
    provider: string;
    amount: string;
    currency: string;
    purpose_kind: string;
    purpose?: Record<string, unknown>;
  }): Promise<CreateOrderResponse> {
    return this.http.request<CreateOrderResponse>("POST", "/v1/payments/orders", {
      body: input,
    });
  }

  async getMyOrder(orderId: string): Promise<PaymentOrder> {
    return this.http.request<PaymentOrder>("GET", `/v1/payments/orders/${orderId}`);
  }

  async listMyOrders(params?: ListParams): Promise<PaymentOrder[]> {
    const res = await this.http.request<{ orders: PaymentOrder[] }>("GET", "/v1/payments/orders", {
      query: listQuery(params),
    });
    return res.orders ?? [];
  }

  async verifyReceipt(input: {
    order_id: string;
    receipt: string;
  }): Promise<{ order: PaymentOrder; transaction_id: string; idempotent_replay?: boolean }> {
    return this.http.request("POST", "/v1/payments/ios/verify-receipt", { body: input });
  }
}
